Enonic UI: Icon image-handling #9077

skjulestad opened the pull request:
Added image rendering to ContentIcon.tsx with fallback for error handling.

skjulestad committed:
Enonic UI: Icon image-handling #9077

Added image rendering to ContentIcon.tsx with fallback for error handling.

skjulestad committed:
Enonic UI: Icon image-handling #9077

Removed spacing between curly-brackets

skjulestad committed:
Enonic UI: Icon image-handling #9077

Changed Vector object fit for more universal sizing.

edloidas commented on modules/lib/src/main/resources/assets/js/app/publish/PublishDialogDependantList.ts:
```diff
@@ -36,11 +39,17 @@ export class PublishDialogDependantList extends DialogDependantItemsList {
+        const summary = content.getContentSummary();
+        if (summary) {
+            viewer.setObject(summary);
+        }
```
This isn’t quite correct. You’re mixing two presentational components here. On top of that, the **Viewer** you’re using is the legacy component we’re actually trying to remove at the moment.

_Think of it like this: you had an old Windows laptop that could play MKV video files Now you’ve switched to a new Mac - modern, sleek, but it doesn’t natively support MKV yet. Instead of installing the right codec, you’re setting up your old laptop next to it just to play those files. That’s clunky and exactly what we should avoid._

We must not use the Viewer here.
What you actually need is just the `resolveIconUrl()` method. It’s relatively small, so you can call the utility class directly: `ContentIconUrlResolver`.

Even better, you could add a method in the `ContentIcon` file that uses `UriHelper` directly, like in the resolver mentioned above:

```js
${UriHelper.trimUrlParams(url)}?${UriHelper.encodeUrlParams(params)}
```

You’ll just need to pass the right parameters like `size` or `crop`.

* `crop` can actually be omitted.
* `size` should be set reasonably - something like double the intended icon size (e.g. `64`) works well.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/CheckableListItemWithStatus.tsx:
```diff
@@ -7,13 +8,16 @@ import {ContentIcon} from './ContentIcon';
+    const imageUrl = summary ? contentViewer.resolveIconUrl(summary) : null;
     const status = CompareStatusFormatter.formatStatusText(content.getCompareStatus());
-    const icon = <ContentIcon contentType={content.getType()}/>;
+    const icon = <ContentIcon contentType={content.getType()} imageUrl={imageUrl}/>;
```
Looking at this, I think it makes more sense to handle the image inside ContentIcon, since it may be used in multiple places. We don’t want to duplicate the URL-resolving logic.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -24,61 +24,136 @@ import {
+    default:
+        return <FileIcon/>;
+    }
+};
```
It can stay here, but I don’t see any reason why it should look like a regular function instead of a component:

```tsx
const FallbackContentIcon = ({type}: {type: ContentTypeName}): JSX.Element => {
```

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -24,61 +24,136 @@ import {
+                src={imageUrl}
+                fallback={fallbackIconFor(contentType)}
+            />
+        );
```
Are you returning image with fallback four times? It might be reasonable just to remove the case block for `ContentTypeName.MEDIA_VECTOR`, `ContentTypeName.MEDIA_IMAGE`, `ContentTypeName.IMAGE` then. Also, for such a small icon size, `object-fit: cover;` might be the universal choice. `contain` might look awful for wide and long images.

skjulestad replied:
We need contain for external icons like author and cover for svg/images as contain is clipping the icons. Svg's / images need cover to look good and square along other icons. All icons are rendered as either svg og img so we do need to separate by type to avoid miss-matching cover and cointain.

Without it we will get either of these end results:
<img width="33" height="38" alt="image" src="https://github.com/user-attachments/assets/fb9a235e-bd63-4a56-96f0-38b62b060a14" />
<img width="138" height="104" alt="image" src="https://github.com/user-attachments/assets/95da9e19-2430-44a3-91a4-7a870c266a85" />

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -24,61 +24,136 @@ import {
+        return fallback;
+    }
+    return <img className={imgClassName} alt={alt} src={src} onError={() => setErrored(true)}/>;
+};
```
Move this to a separate file and name it simply `<Image>`. This might become our default way of rendering images. There could be a lot going on here, but let’s keep it simple for now:

```tsx
<img
  className={className}
  alt={alt}
  width={size}        // optional
  height={size}       // optional
  loading={lazy ? 'lazy' : 'eager'} // eager by default
  decoding="async"
  referrerPolicy="no-referrer"
  src={src}
  srcSet={srcSet}     // omit for now, but we may add it later
  onError={handleError}
/>
```

`src` must always be set and treated as non-null.

skjulestad committed:
Enonic UI: Icon image-handling #9077

Moved image-handling inside ContentIcon.tsx
Readjusted fallback to avoid duplication
Created reusable image component.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/CheckableListItemWithStatus.tsx:
```diff
@@ -1,33 +1,41 @@
+    const icon = (
+        <ContentIcon
+            contentType={content.getType()}
+            summary={content.getContentSummary?.()}
```
I'm sure that we don't need optional chaining here, as `getContentSummary` function is present in objects of this type. You don't need to verify it exists.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -1,84 +1,94 @@
+    if (crop) {
+        params.crop = true;
+    }
+    return `${UriHelper.trimUrlParams(url)}?${UriHelper.encodeUrlParams(params)}`;
```
Why are you using both `ContentIconUrlResolver` and `UriHelper` here? I see that the first line there is doing exactly the same as the rest of the code here, but inside the `ContentIconUrlResolver` class. 
Ideally, we must create a utility function in a new file with no defaults and use it here. Just a smaller and smarter version of `ContentIconUrlResolver`:
```ts
interface ImageUrlParams {
    size?: number | null;
    crop?: boolean | null;
}

export function createImageUrl(url: string, params: ImageUrlParams): string {
    return `${UriHelper.trimUrlParams(url)}?${UriHelper.encodeUrlParams(params)}`;
}
```

skjulestad committed:
Enonic UI: Icon image-handling #9077

Updated styling to fit better with Figma design.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -1,84 +1,94 @@
+    }
+
+    return tryImage() ?? <FileIcon/>;
+};
```
While it's generally correct, it's hard to read, as it mixes variables, function call and components.
Also, regarding rendering logic there:
- If the content is image, you try image -> icon -> FileIcon (obsolete, but done for type safety, as your map has string keys)
- You try icons instead with fallback to image, if type is unknown. File won't probably be used here also.
To simplify to simplify the flow, we can write it like this:
```tsx
if (isImageType) {
    return tryImage(imgClassName) ?? icon ?? <FileIcon />;
}

return icon ?? tryImage() ?? icon ?? <FileIcon />;
```
I think that `imgClassName` should not be called like this. Actually, class must be set here, in this component and not from the outside.
So, we use image, if content type is unknown, or if it's image type.

edloidas committed:
Enonic UI: Icon image-handling #9077

Updated ESLint config for TSX files to force types.
Moved url resolving logic to a separate utility file.
Simplified icon and image components.

skjulestad committed:
Enonic UI: Icon image-handling #9077
